streams/shared/src/main/scala/zio/stream/ZChannel.scala

```diff
@@ -722,52 +727,55 @@ sealed trait ZChannel[-Env, -InErr, -InElem, -InDone, +OutErr, +OutElem, +OutDon
         queueReader  = ZChannel.fromInput(input)
         outgoing    <- Queue.bounded[Exit[Either[Unit, OutDone], OutElem2]](bufferSize)
         _           <- scope.addFinalizer(outgoing.shutdown)
-        errorSignal <- Promise.make[Unit, Nothing]
+        errorSignal <- Promise.make[Nothing, Unit]
         permits     <- Semaphore.make(n.toLong)
         failure     <- Ref.make[Cause[OutErr1]](Cause.empty)
-        pull        <- (queueReader >>> self).toPullIn(scope)
+        pull        <- (queueReader >>> self).toPullInAlt(scope)
         _ <- pull
-               .foldCauseZIO(
-                 cause =>
-                   failure.update(_ && cause) *>
-                     outgoing.offer(Exit.fail(Left(()))) *>
-                     Exit.failCause(Cause.unit),
-                 {
-                   case Left(outDone) =>
-                     permits.withPermits(n.toLong)(ZIO.unit).interruptible *> outgoing.offer(Exit.fail(Right(outDone)))
-                   case Right(outElem) =>
-                     for {
-                       latch <- Promise.make[Nothing, Unit]
-                       _ <- permits.withPermit {
-                              latch.succeed(()) *>
-                                ZIO.uninterruptibleMask { restore =>
-                                  restore(errorSignal.await).raceFirstAwait(
-                                    restore(f(outElem))
-                                      .catchAllCause(cause => failure.update(_ && cause) *> Exit.failCause(Cause.unit))
-                                  )
-                                }.foldCauseZIO(
-                                  _ => outgoing.offer(Exit.fail(Left(()))) *> errorSignal.refailCause(Cause.unit),
-                                  elem => outgoing.offer(Exit.succeed(elem))
-                                )
-                            }.forkIn(scope)
-                       _ <- latch.await
-                     } yield ()
-                 }
+               .flatMap(outElem =>
+                 for {
+                   latch <- Promise.make[Nothing, Unit]
+                   _ <- permits
+                          .withPermit(latch.succeed(()) *> f(outElem))
+                          .flatMap(elem => outgoing.offer(Exit.succeed(elem)))
+                          .catchAllCause(cause =>
+                            failure.update(_ && cause).unless(cause.isInterrupted) *> errorSignal.succeed(
+                              ()
+                            ) *> outgoing.offer(Exit.fail(Left(())))
+                          )
+                          .fork
```

Comment:
Why remove the `forkIn`? I guess we don't need to await the fiber due to the latch?

Reply:
The worker fibers will be interrupted as soon as the consumer fiber terminates, which happens when the error signal kicks in